File: 0332-reconstruct-itinerary/0332-reconstruct-itinerary.go
type Graph map[string][]string
func findItinerary(tickets [][]string) []string {
    graph := buildGraph(tickets)
    result := make([]string, 0, len(tickets)+1)
    dfs(graph, "JFK", &result)
    reverse(result)
    return result
}
func buildGraph(tickets [][]string)Graph{
    graph := make (Graph)
    for _, ticket := range  tickets{
        from, to := ticket[0], ticket[1]
        graph[from] =  append(graph[from], to)
    }
    for from := range graph{
        sort.Strings(graph[from])
    }
    return graph
}

func dfs(graph Graph, airport string, result *[]string){
    for len(graph[airport])>0{
        next := graph[airport][0]
        graph[airport] = graph[airport][1:]
        dfs(graph, next, result)
    }
    *result = append(*result, airport)
}
func reverse(s [] string){
    for i, j := 0, len(s)-1 ; i<j ; i, j = i+1,j-1{
        s[i], s[j] = s[j], s[i]
    } 
}

File: 0332-reconstruct-itinerary/0332-reconstruct-itinerary.rs
use std::collections::{HashMap, BinaryHeap};
use std::cmp::Reverse;

type Graph = HashMap<String, BinaryHeap<Reverse<String>>>;

impl Solution {
    pub fn find_itinerary(tickets: Vec<Vec<String>>) -> Vec<String> {
        let mut graph = Self::build_graph(tickets);
        let mut route = Vec::new();

        Self::dfs("JFK", &mut graph, &mut route);
        route.reverse();
        route
    }
    fn build_graph(tickets : Vec<Vec<String>>) -> Graph{
        let mut graph = HashMap::new();
        for ticket in tickets{
            graph.entry(ticket[0].clone())
                .or_insert(BinaryHeap::new())
                .push(Reverse(ticket[1].clone()));
        }
        graph
    }
    fn dfs(airport : &str , graph : &mut Graph, route: &mut Vec<String>){
        while let Some(dest) = graph.get_mut(airport).and_then(|x| x.pop()){
            Self::dfs(&dest.0,  graph, route);
        }
        route.push(airport.to_string());
    }
}